[CK_BUILDER]ckb add remining fwd conv device ops

JH-Leon-KIM-AMD opened the pull request:
## Proposed changes

This PR adds factory support for the remaining forward convolution device operations in CK Builder.

**Jira Ticket:** https://amd-hub.atlassian.net/jira/software/c/projects/ALMIOPEN/boards/319/backlog?selectedIssue=ALMIOPEN-350

**Task 350** - Add remaining forward convolution device operations:

1. **DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK** 
   - Added DL (Direct Load) factory specialization for NHWC layout-specific convolution
   - Created new DL-specific algorithm descriptor with 30 template parameters
   - Added test helper function and 3 test cases covering DEFAULT and FILTER_1X1_PAD0 specializations
   - All tests passing (15/15 builder tests)

2. **DeviceGroupedConvFwdMultipleD_Xdl_CShuffle_Large_Tensor** 
   - Added Large_Tensor factory specialization for N-dimension splitting (large-than-memory tensors)
   - Implemented macro collision workaround using pragma push/pop for GridwiseGemmTemplateParameters
   - Reuses existing XDL algorithm descriptor (42 identical template parameters)
   - Added test helper function and 2 test cases covering DEFAULT and FILTER_1X1_PAD0 specializations
   - All tests passing (15/15 builder tests)

This completes Task 350 - all 4 forward convolution device operations are now supported in CK Builder:
- ✅ DeviceGroupedConvFwdMultipleABD_Xdl_CShuffle (by Ville)
- ✅ DeviceGroupedConvFwdMultipleD_Wmma_CShuffle (by Ville)
- ✅ DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK (this PR)
- ✅ DeviceGroupedConvFwdMultipleD_Xdl_CShuffle_Large_Tensor (this PR)

## Checklist

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

### Design Decisions

**1. DL Algorithm Descriptor (30 parameters)**
- DL uses VALU instructions instead of XDL matrix cores, requiring different parameter structure
- Fixed NHWC_KYXC_NHWK layout only (not flexible like XDL)
- Created separate descriptor type: `ConvAlgorithm_DeviceGroupedConvFwdDlMultipleD_NHWC_KYXC_NHWK`
- New DL-specific concepts: `DlThreadConfigDescriptor`, `DlThreadClusterDescriptor`, `DlBlockTransferK0M0M1K1Descriptor`, etc.

**2. Large_Tensor Descriptor Reuse (42 parameters)**
- Large_Tensor has identical template parameters to regular XDL CShuffle
- Only difference: internal `SplitN=true` flag in device operation (not exposed in factory interface)
- Reuses existing descriptor: `ConvAlgorithm_DeviceGroupedConvFwdMultipleABD_Xdl_CShuffle`
- No new descriptor or concepts needed

**3. Macro Collision Workaround**
- Both `device_grouped_conv_fwd_multiple_abd_xdl_cshuffle.hpp` (line 41) and `device_grouped_conv_fwd_multiple_d_xdl_large_tensor_cshuffle.hpp` (line 51) define `GridwiseGemmTemplateParameters` macro without `#undef`
- Used `#pragma push_macro`/`#pragma pop_macro` to isolate Large_Tensor header's macro scope
- This may need to fix on CK headers

vpietila-amd committed:
Add device operation to conv signature. Use unions to hold conv layouts and device operations.

vpietila-amd committed:
Add predicates for all device op instances.

vpietila-amd committed:
Use the device op signature for validation.

vpietila-amd committed:
Merge branch 'develop' into vpietila/ckb-generalize-conv-factory

vpietila-amd committed:
Fix ckb CMakeLists.txt file for tests.

vpietila-amd committed:
Fix building CK Builder instance traits after the introduction of direct load template parameter in CK.

vpietila-amd committed:
Merge branch 'vpietila/fix-building-ckb-tests' into vpietila/ckb-generalize-conv-factory

vpietila-amd committed:
Fix clang-formatting.

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into jeonghyun/ckb-add-remining-fwd-conv-device-ops

JH-Leon-KIM-AMD committed:
add device_grouped_conv_fwd_dl_multiple_d_nhwc_kyxc_nhwk

JH-Leon-KIM-AMD committed:
Add full DL configurability with Option A implementation

- Added 5 DL descriptor structs (39 configurable parameters)
- Added 10 C++20 concepts for type-safe validation
- Updated factory to read all parameters from descriptors
- Updated test helper to populate all descriptors
- All tests passing (13/13 including 3 new DL tests)

JH-Leon-KIM-AMD committed:
Add factory and test support for DeviceGroupedConvFwdMultipleD_Xdl_CShuffle_Large_Tensor

- Add factory specialization for Large_Tensor device operation (conv_factory.hpp lines 1145-1265)
- Add macro collision workaround using pragma push/pop (conv_factory.hpp lines 43-51)
- Add test helper function run_test_DeviceGroupedConvFwdMultipleD_Xdl_CShuffle_Large_Tensor
- Add builder test file test_ckb_conv_fwd_2d_large_tensor_fp16.cpp with 2 test cases
- Update CMakeLists.txt to include new test file
- Reuse existing ConvAlgorithm_DeviceGroupedConvFwdMultipleABD_Xdl_CShuffle descriptor
- Map all 42 template parameters identical to regular XDL CShuffle
- All 15 builder tests passing including 2 new Large_Tensor tests

Completes Task 350: All 4 forward convolution device operations now supported in CK Builder.

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into jeonghyun/ckb-add-remining-fwd-conv-device-ops

vpietila-amd commented on experimental/builder/test/conv/test_ckb_conv_fwd_2d_dl_fp16.cpp:
```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2025, Advanced Micro Devices, Inc. All rights reserved.
```
Check that the Copyright follows the convention introduced in PR #3150. Same for the other new test file.

JH-Leon-KIM-AMD replied:
Thank you!

vpietila-amd reviewed (dismissed):
Looks good to me. There's a PR #3154 that will reduce the code duplication in the test code. If that PR goes in first, we need to do the same changes for this PR too. But they are relatively straightforward changes. Once this PR is merged in the `develop` branch, I'll do another refactoring to remove the explicit device op flag from the signature.

JH-Leon-KIM-AMD committed:
Update copyright headers to new format

- Change copyright format to: Copyright (C) Advanced Micro Devices, Inc., or its affiliates.
- Reorder headers: Copyright first, then SPDX-License-Identifier
- Updated files:
  * experimental/builder/test/conv/test_ckb_conv_fwd_2d_dl_fp16.cpp
  * experimental/builder/test/conv/test_ckb_conv_fwd_2d_large_tensor_fp16.cpp
  * experimental/builder/include/ck_tile/builder/device_op_types.hpp

JH-Leon-KIM-AMD committed:
Merge branch 'develop' into jeonghyun/ckb-add-remining-fwd-conv-device-ops

JH-Leon-KIM-AMD committed:
fix c++ 18 format

JH-Leon-KIM-AMD committed:
Fix clang-format-18 error in device_op_types.hpp
